### PR TITLE
consenus: Implement LLMQ_100_67 quorums

### PR DESCRIPTION
**Pull request by UdjinM6**
Implements #3729

Re-uses DEPLOYMENT_V17 as an activation mechanism.

### CODE REVIEW

**Comment by xdustinface** [comment]
See [suggestions-3844](https://github.com/xdustinface/dash/tree/suggestions-3844) for some renaming ideas. Feel free to pick only what you like 😄 It's nothing which would hold me up giving an utACK. 

My thoughts about the `llmq_type_new` to `llmq_type_v17`: If we add another one its not the new one anymore haha.

**Comment by UdjinM6** [comment]
Picked some of your suggestions. How about de92158 instead of the rest of them?

**Review by xdustinface — COMMENTED** [review]
> How about [de92158](https://github.com/dashpay/dash/commit/de92158df3151d3011baee0f3e51502b1f765b45) instead of the rest of them?

Okay yeah agree, the term "new" might make sense there if we add another LLMQ at some point :D

There is one thing though where im not totally sure about yet: Will things work well if we have all members of 100 member LLMQs connected (Spork21 is currently active for this new LLMQ Type)?

@codablock You might have tested that? Can you share your thoughts here also?

File: src/rpc/rpcquorums.cpp [hunk 1]
```diff
@@ -54,7 +54,7 @@ UniValue quorum_list(const JSONRPCRequest& request)
 
     UniValue ret(UniValue::VOBJ);
 
-    for (auto& p : Params().GetConsensus().llmqs) {
+    for (auto& p : llmq::CLLMQUtils::GetEnabledQuorumTypes(chainActive.Tip())) {
```
Comment (PastaPastaPasta) [inline src/rpc/rpcquorums.cpp:57]: Would it not be the tip before this? (I actually think not, but want to through it out there)

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

File: src/llmq/quorums_utils.cpp [hunk 2]
```diff
@@ -254,5 +257,42 @@ bool CLLMQUtils::IsQuorumActive(Consensus::LLMQType llmqType, const uint256& quo
...
+        case Consensus::LLMQ_DEVNET:
+            break;
+        default:
+            throw std::runtime_error(strprintf("%s: Unknown LLMQ type %d", __func__, llmqType));
+    }
+
+    return true;
+}
+
+std::map<Consensus::LLMQType, Consensus::LLMQParams> CLLMQUtils::GetEnabledQuorumTypes(const CBlockIndex* pindex)
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/llmq/quorums_utils.cpp]: Picked the first commit only. The second one looks unrelated and should be in a separate PR imo.

File: src/rpc/rpcquorums.cpp (hunk 1 above)
Comment (UdjinM6) [reply src/rpc/rpcquorums.cpp]: Basically we should ensure data consistency in different parts of code e.g. it's `chainActive.Tip()` in `CreateNewBlock` and `pindex->pprev` in `CQuorumBlockProcessor::ProcessBlock`. For RPCs it's kind of not _that_ critical but IMO it makes sense to use chain tip here to see the same data miners do.

File: src/llmq/quorums_utils.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/llmq/quorums_utils.cpp]: Agreed!

File: src/rpc/rpcquorums.cpp (hunk 1 above)
Comment (PastaPastaPasta) [reply src/rpc/rpcquorums.cpp]: Okay, I think that makes sense

**Comment by PastaPastaPasta** [comment]
Just talked with @Quantumexplorer, due to modifications he is making in tenderdash, there is no longer a need for a high minimum size of 90 (no longer using a dynamic threshold), so I propose we drop that to 80.

File: src/chainparams.cpp [hunk 3]
```diff
@@ -262,6 +282,26 @@ static Consensus::LLMQParams llmq400_85 = {
...
+        .dkgInterval = 24, // one DKG per hour
+        .dkgPhaseBlocks = 2,
+        .dkgMiningWindowStart = 10, // dkgPhaseBlocks * 5 = after finalization
+        .dkgMiningWindowEnd = 18,
+        .dkgBadVotesThreshold = 80,
+
+        .signingActiveQuorumCount = 24, // a full day worth of LLMQs
+
+        .keepOldConnections = 25,
+        .recoveryMembers = 50,
```
[the remark this answers is not in the document]
Comment (thephez) [reply src/chainparams.cpp:302]: Per https://github.com/dashpay/dash/blob/master/src/consensus/params.h#L119-L121, `keepOldConnections` has to be 25 (or more).

File: src/chainparams.cpp [hunk 4]
```diff
@@ -262,6 +282,26 @@ static Consensus::LLMQParams llmq400_85 = {
         .recoveryMembers = 100,
 };
 
+// Used for Platform
+static Consensus::LLMQParams llmq100_67 = {
+        .type = Consensus::LLMQ_100_67,
+        .name = "llmq_100_67",
+        .size = 100,
+        .minSize = 90,
```
Comment (PastaPastaPasta) [inline src/chainparams.cpp:290]:
```suggestion
        .minSize = 80,
```

File: src/llmq/quorums_utils.cpp [hunk 5]
```diff
@@ -254,5 +257,47 @@ bool CLLMQUtils::IsQuorumActive(Consensus::LLMQType llmqType, const uint256& quo
     return false;
 }
 
+bool CLLMQUtils::IsQuorumTypeEnabled(Consensus::LLMQType llmqType, const CBlockIndex* pindex)
+{
+    const Consensus::Params& consensusParams = Params().GetConsensus();
+    bool f_v17_Active =  VersionBitsState(pindex, consensusParams, Consensus::DEPLOYMENT_V17, versionbitscache) == ThresholdState::ACTIVE;
```
Comment (PastaPastaPasta) [inline src/llmq/quorums_utils.cpp:263]: does it make sense to combine these lines?

```suggestion
    bool f_v17_Active =  VersionBitsState(pindex, Params().GetConsensus(), Consensus::DEPLOYMENT_V17, versionbitscache) == ThresholdState::ACTIVE;
```

Comment (UdjinM6) [reply src/llmq/quorums_utils.cpp:263]: Imo current version is slightly more readble

**Comment by PastaPastaPasta** [comment]
What is the behavior of this quorum in regards to Concentrated Recovery?

**Comment by UdjinM6** [comment]
There are no changes in `CLLMQUtils::IsAllMembersConnectedEnabled` atm so it's going to participate just like 50_60 quorum does. Should I exclude it? (cant find anything about it in the original issue)

**Comment by PastaPastaPasta** [comment]
It wasn't mentioned in the original issue... I'm not sure if it's a good idea or not. I guess it would depend on if concentrated recovery / MNs could handle another massive hunk of connections...

File: test/functional/feature_new_quorum_type_activation.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python3
+# Copyright (c) 2015-2020 The Dash Core developers
```
Comment (PastaPastaPasta) [inline test/functional/feature_new_quorum_type_activation.py:2]:
```suggestion
# Copyright (c) 2020 The Dash Core developers
```

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

**Review by PastaPastaPasta — APPROVED** [review]
LGTM, utACK, at this point waiting on tests

**Review by xdustinface — APPROVED** [review]
utACK

**Pull request merged**